contracts/InterchainProposalSender.sol

```diff
@@ -44,154 +45,87 @@ contract InterchainProposalSender is IProposalSender {
-        uint fee,
-        address[] memory targets,
-        uint256[] memory values,
-        string[] memory signatures,
-        bytes[] memory data
+    function _sendProposal(
+        InterchainCalls.InterchainCall memory xCall
     ) internal {
-        revertIfInvalidProposalArgs(targets, values, signatures, data);
-
-        if (fee == 0) {
+        if (xCall.gas == 0) {
             revert InvalidFee();
         }
 
-        bytes memory payload = abi.encode(
-            msg.sender,
-            targets,
-            values,
-            signatures,
-            data
-        );
+        bytes memory payload = abi.encode(msg.sender, xCall.calls);
 
-        gasService.payNativeGasForContractCall{value: fee}(
+        gasService.payNativeGasForContractCall{value: xCall.gas}(
             address(this),
-            destinationChain,
-            destinationContract,
+            xCall.destinationChain,
+            xCall.destinationContract,
             payload,
             msg.sender
         );
 
-        gateway.callContract(destinationChain, destinationContract, payload);
-    }
-
-    function revertIfInvalidProposalArgs(
-        address[] memory targets,
-        uint256[] memory values,
-        string[] memory signatures,
-        bytes[] memory data
-    ) private pure {
-        if (
-            targets.length == 0 ||
-            targets.length != values.length ||
-            targets.length != signatures.length ||
-            targets.length != data.length
-        ) {
-            revert ProposalArgsMisMatched();
-        }
-    }
-
-    function revertIfInvalidLength(
-        string[] memory destinationChains,
-        string[] memory destinationContracts,
-        uint[] memory fees,
-        address[][] memory targets,
-        uint256[][] memory values,
-        string[][] memory signatures,
-        bytes[][] memory data
-    ) private pure {
-        if (
-            destinationChains.length != destinationContracts.length ||
-            destinationChains.length != fees.length ||
-            destinationChains.length != targets.length ||
-            destinationChains.length != values.length ||
-            destinationChains.length != signatures.length ||
-            destinationChains.length != data.length
-        ) {
-            revert ArgsLengthMisMatched();
-        }
+        gateway.callContract(
+            xCall.destinationChain,
+            xCall.destinationContract,
+            payload
+        );
     }
 
-    function revertIfInvalidFee(uint[] memory fees) private {
+    function revertIfInvalidFee(
+        InterchainCalls.InterchainCall[] calldata xCalls
+    ) private {
         uint totalFees = 0;
-        for (uint i = 0; i < fees.length; i++) {
-            totalFees += fees[i];
+        for (uint i = 0; i < xCalls.length; ) {
+            totalFees += xCalls[i].gas;
+            unchecked {
+                ++i;
+            }
         }
 
         if (totalFees != msg.value) {
```

Comment:
if total fees is more than the balance, the call will revert anyways. is it useful to ensure it matches msg.value?